beluga/include/beluga/random/uniform_grid_distribution.hpp

```diff
@@ -0,0 +1,81 @@
+// Copyright 2024 Ekumen, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BELUGA_RANDOM_UNIFORM_GRID_DISTRIBUTION_HPP
+#define BELUGA_RANDOM_UNIFORM_GRID_DISTRIBUTION_HPP
+
+#include <random>
+
+#include <range/v3/utility/random.hpp>
+#include <sophus/se2.hpp>
+
+#include <beluga/sensor/data/occupancy_grid.hpp>
+
+namespace beluga {
+
+/// Primary template for a uniform grid distribution.
+/**
+ * \tparam T The result type for sampling from the distribution.
+ */
+template <class T>
+class UniformGridDistribution;
+
+/// Specialization of uniform grid distribution for Sophus::SE2d.
+template <>
+class UniformGridDistribution<Sophus::SE2d> {
+ public:
+  /// Constructs a uniform grid distribution based on the provided occupancy grid.
+  /**
+   * \tparam Derived A type derived from BaseOccupancyGrid2.
+   * \param grid The occupancy grid from which free states will be computed.
+   */
+  template <class Derived>
+  constexpr explicit UniformGridDistribution(const BaseOccupancyGrid2<Derived>& grid)
+      : free_states_{compute_free_states(static_cast<const Derived&>(grid))},
+        distribution_{0, free_states_.size() - 1} {
+    assert(free_states_.size() > 0);
+  }
+
+  /// Generates a random 2D pose.
+  /**
+   * This function generates a random pose by sampling a random rotation
+   * from SO2 space and a random translation from the precomputed free states
+   * based on the provided occupancy grid.
+   *
+   * \tparam URNG The type of the random number generator.
+   * \param engine The random number generator engine (default is a default generator).
+   * \return A random Sophus::SE2d pose.
+   */
+  template <class URNG = typename ranges::detail::default_random_engine>
+  [[nodiscard]] Sophus::SE2d operator()(URNG& engine = ranges::detail::get_random_engine()) {
```

Comment:
@nahueespinosa hmm, why provide a default URNG? STL distributions don't, and we already have `beluga::views::sample` to inject the ranges-v3 specific URNG.

Reply:
Also, could the distribution yield particles instead of particle state? Perhaps it's two distributions, one composing the other.

Reply:
> STL distributions don't, and we already have beluga::views::sample to inject the ranges-v3 specific URNG.

That's fair.

> Also, could the distribution yield particles instead of particle state?

It can, but it's not consistent with `MultivariateNormalDistribution` for instance. In my mind, distributions generate states that we use to initialize particles, there is no need to make that relationship stronger than that in my opinion.

> Perhaps it's two distributions, one composing the other.

Could you elaborate on that?

Reply:
See https://github.com/Ekumen-OS/beluga/pull/319/commits/85f562321b322545c73cde9256e91784de745b6d.